smart_card_connector_app/src/chrome-api-provider.js

```diff
@@ -0,0 +1,165 @@
+ * which correspond to different PCSC requests. When an event is fired,
+ * ChromeApiProvider sends a request to Pcsc.ServerRequestHandler and waits for
+ * the response. When the response is received, it reports the result back using
+ * chrome.smartCardProviderPrivate API.
+ */
+GSC.ConnectorApp.ChromeApiProvider = class extends goog.Disposable {
+  /**
+   * @param {!goog.messaging.AbstractChannel} serverMessageChannel
+   * @param {!Pcsc.ReadinessTracker} serverReadinessTracker
+   */
+  constructor(serverMessageChannel, serverReadinessTracker) {
+    super();
+    this.serverRequester_ = new Pcsc.ServerRequestHandler(
+        serverMessageChannel, serverReadinessTracker, 'chrome');
+
+    this.chromeEventListeners_ = [];
+    this.subscribeToChromeApiEvents_();
+
+    this.serverRequester_.addOnDisposeCallback(() => this.dispose());
+  }
+
+  /** @override */
+  disposeInternal() {
+    this.unsubscribeFromChromeApiEvents_();
+    this.serverRequester_.dispose();
+    super.disposeInternal();
+  }
+
+  /**
+   * Subscribes to chrome.smartCardProviderPrivate events.
+   * @private
+   */
+  subscribeToChromeApiEvents_() {
+    if (chrome.smartCardProviderPrivate === undefined) {
+      goog.log.warning(
+          logger,
+          'chrome.smartCardProviderPrivate API is not available, ' +
+              'could not subscribe to the API events.');
+      return;
+    }
+
+    this.chromeEventListeners_.push(new ChromeEventListener(
+        chrome.smartCardProviderPrivate.onEstablishContextRequested,
+        (requestId) => this.establishContextListener_(requestId)));
+  }
+  /**
+   * Unsubscribes from chrome.smartCardProviderPrivate events.
+   * @private
+   */
+  unsubscribeFromChromeApiEvents_() {
+    this.chromeEventListeners_.forEach(listener => listener.dispose());
+  }
+
+  /**
+   * @param {number} requestId
+   * @private
+   */
+  async establishContextListener_(requestId) {
+    const callArguments = [PcscApi.SCARD_SCOPE_SYSTEM, null, null];
+    const remoteCallMessage =
+        new GSC.RemoteCallMessage('SCardEstablishContext', callArguments);
+
+    var resultCode = chrome.smartCardProviderPrivate.ResultCode.INTERNAL_ERROR;
+    var sCardContext = 0;
+
+    try {
+      const responseItems =
+          await this.serverRequester_.handleRequest(remoteCallMessage);
+
+      const result = new PcscApi.SCardEstablishContextResult(responseItems);
+      result.get(
+          (context) => {
+            sCardContext = context;
+            resultCode = chrome.smartCardProviderPrivate.ResultCode.SUCCESS;
+          },
+          (errorCode) => {
+            resultCode = convertErrorCodeToEnum(errorCode);
+          });
+    } catch (error) {
+      goog.log.log(
+          logger,
+          this.isDisposed() ? goog.log.Level.FINE : goog.log.Level.WARNING,
+          `Failed to process onEstablishContextRequested event
+           with request id ${requestId}: ${error}`);
```

Comment:
nit: I believe the line break and the indentation will surface in the log messages in a confusing way. One workaround would be to use concatenation:

```
          `Failed to ... ` +
              `foo bar`
```

Reply:
Fixed. But actually I checked this template literal version: in console logs for extension it was printing the message without line breaks (just with a single whitespace instead of line break). I'm not sure how that works since the doc says it preserves indentation ¯\_(ツ)_/¯